### PR TITLE
RStudio IE and IE Framework Enhancements

### PR DESCRIPTION
**Pull request by hexylena**
Includes the entire history of the RStudio IE.

Depends on
- [x]  merging #229 **first**.
- [x] #239

### CODE REVIEW

**Comment by jmchilton** [comment]
Can we label this one WIP also? We probably should add an RData datatype as part of this.

**Comment by hexylena** [comment]
@jmchilton sure, that's fine.

File: config/plugins/interactive_environments/rstudio/templates/rstudio.mako [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+## General IE specific
+# Access URLs for the notebook from within galaxy.
+# TODO: Make this work without pointing directly to IE. Currently does not work
+# through proxy.
+notebook_pubkey_url = ie_request.url_template('${PROXY_URL}/rstudio/${PORT}/auth-public-key')
+notebook_access_url = ie_request.url_template('${PROXY_URL}/rstudio/${PORT}/')
+notebook_login_url =  ie_request.url_template('${PROXY_URL}/rstudio/${PORT}/auth-do-sign-in')
+
+# Did the user give us an RData file?
+if hda.datatype.__class__.__name__ == "RData":
```
Comment (jmchilton) [inline config/plugins/interactive_environments/rstudio/templates/rstudio.mako:24]: What do you think about subtyping RData for this purpose (RStudio IE workspace)? Does that make sense - is there special stuff in the environment - or is one workspace the same as another?

Comment (hexylena) [inline config/plugins/interactive_environments/rstudio/templates/rstudio.mako:24]: Huh. That's actually a good question, I hadn't given that any thought.

`.RData` files should just be .gz of data structures loaded in the R session. I don't know think anything is included in the IE that wouldn't be able to be used outside of the IE, so I'd lean towards keeping it as-is, just raw RData files. I'll double check this tomorrow, that the .RData files are completely functional inside and out. If they are, I'd say no reason to impose any extraneous limits.

Comment (hexylena) [inline config/plugins/interactive_environments/rstudio/templates/rstudio.mako:24]: Okay, doesn't look like there's a need to subclass anything. Let's just go with this as-is, whenever you're happy with the PR.

File: config/plugins/interactive_environments/rstudio/config/rstudio.ini.sample [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+# This cannot be changed
+password_auth = True
+# Other
+apache_urls = False
+ssl = False
+
+[docker]
+command = docker
+image = erasche/docker-rstudio-notebook:dev
+
```
Comment (jmchilton) [inline config/plugins/interactive_environments/rstudio/config/rstudio.ini.sample:11]: @erasche I don't think this tag exists (https://registry.hub.docker.com/u/erasche/docker-rstudio-notebook/tags/manage/) - it seems to for IPython (https://registry.hub.docker.com/u/bgruening/docker-ipython-notebook/tags/manage/). Can you mimic @bgruening's IPython setup or switch this to `:latest`?

Comment (hexylena) [inline config/plugins/interactive_environments/rstudio/config/rstudio.ini.sample:11]: My intention was to mirror @bgruening's. Whoops. Apparently I didn't turn on the `:dev` tag. Image should be building now. **Thank you** for catching that.

Comment (hexylena) [inline config/plugins/interactive_environments/rstudio/config/rstudio.ini.sample:11]: https://registry.hub.docker.com/u/erasche/docker-rstudio-notebook/builds_history/41411/

**Comment by jmchilton** [comment]
So #229 is listed as prerequisite but this PR uses `write_conf_file` which is renamed as part of that.

```
galaxy.webapps.galaxy.controllers.visualization ERROR 2015-05-11 02:33:55,781 error rendering visualization (rstudio): 'InteractiveEnviornmentRequest' object has no attribute 'write_conf_file'
Traceback (most recent call last):
  File "/opt/galaxy/galaxy-app/lib/galaxy/webapps/galaxy/controllers/visualization.py", line 732, in render
    return plugin.render( trans=trans, embedded=embedded, **kwargs )
  File "/opt/galaxy/galaxy-app/lib/galaxy/visualization/plugins/plugin.py", line 139, in render
    return self._render( render_vars, trans=trans, embedded=embedded )
  File "/opt/galaxy/galaxy-app/lib/galaxy/visualization/plugins/plugin.py", line 280, in _render
    return trans.fill_template( template_filename, template_lookup=self.template_lookup, **render_vars )
  File "/opt/galaxy/galaxy-app/lib/galaxy/web/framework/webapp.py", line 810, in fill_template
    return self.fill_template_mako( filename, **kwargs )
  File "/opt/galaxy/galaxy-app/lib/galaxy/web/framework/webapp.py", line 825, in fill_template_mako
    return template.render( **data )
  File "/opt/galaxy/galaxy-app/eggs/Mako-0.4.1-py2.7.egg/mako/template.py", line 296, in render
    return runtime._render(self, self.callable_, args, data)
  File "/opt/galaxy/galaxy-app/eggs/Mako-0.4.1-py2.7.egg/mako/runtime.py", line 660, in _render
    **_kwargs_for_callable(callable_, data))
  File "/opt/galaxy/galaxy-app/eggs/Mako-0.4.1-py2.7.egg/mako/runtime.py", line 692, in _render_context
    _exec_template(inherit, lclcontext, args=args, kwargs=kwargs)
  File "/opt/galaxy/galaxy-app/eggs/Mako-0.4.1-py2.7.egg/mako/runtime.py", line 718, in _exec_template
    callable_(context, *args, **kwargs)
  File "/opt/galaxy/galaxy-app/database/template_cache/rstudio.mako.py", line 60, in render_body
    ie_request.write_conf_file(temp_dir, {'notebook_username': 'galaxy', 'notebook_password': PASSWORD,
AttributeError: 'InteractiveEnviornmentRequest' object has no attribute 'write_conf_file'
```

Thoughts?

**Comment by hexylena** [comment]
This shjouldn't be using write_conf_file, crapsticks. I switched everything in RStudio over to ENV... I must not have pushed that commit. So sorry @jmchilton!

**Comment by hexylena** [comment]
@jmchilton pushed the commits. Sorry about that. Doing some more testing now just in case I missed anything...

**Comment by hexylena** [comment]
@jmchilton okay, this was a less tested PR and definitely more WIP (mostly because I haven't quite figured out the waltz to developing on one branch while using commits from another that don't get included in the history), so I'm going to say "this is still WIP" and I'll get everything 100% tomorrow morning. Sorry about the bugs, thanks again for the code review and sanity checks, I really appreciate it!

**Comment by jmchilton** [comment]
No problem at all - thank you for the fixes and for working this - this is really exciting functionality!

If you are going to work on this more - let me know when you think it is ready to go - I'll work on testing #229 by itself instead in the mean time.

**Comment by hexylena** [comment]
Okay, I'll ping you again when it's out of WIP.

**Comment by hexylena** [comment]
@jmchilton no longer WIP, if you'd like to test it :)

**Comment by jmchilton** [comment]
Wonderful - works for me now! Thanks a million @erasche!

**Comment by jmchilton** [comment]
Mind if I update the title to reflect that there are IE framework enhancements in there also?

**Comment by hexylena** [comment]
Go for it!

**Comment by bgruening** [comment]
Release time! This is exciting!

**Pull request merged**